### PR TITLE
Handle location of player if was moved for not valid world

### LINKED ISSUE #9081: Add `PlayerSpawnInUnknownWorldEvent`
### Is your feature request related to a problem?

Yea, our players sometimes gets stuck in blocks (in main overworld at same position where they were in the other world) and then die due to suffocation :/

This happens only if the server was restarted and the other world's folder (created by plugin or vanilla nether/the end) was deleted.

### Describe the solution you'd like.

Add `PlayerSpawnInUnknownWorldEvent` (or something with better name) that gets called if player joining the game and the server can't find the world this player was in before.

That event contains `#setSpawnLocation(Location)` and nullable `#getSpawnLocation()`. By default that `Location` is `null` to keep existing behavior. If no one plugin changed this `Location`, then this player will spawn in overworld at same position (current behavior), otherwise the server will use that `Location` to spawn this player at.

### Describe alternatives you've considered.

idk. I'm tried `PlayerSpawnLocationEvent` but it gets called with default world/overworld in this case, so plugins cannot determine exactly which world the player was in before.

### Other

_No response_

### PR DESCRIPTION
**Pull request by Doc94**
This clos e #9081 like the issue said when the world of spawn change because the world is unloaded or not exists keep the previous location causing issues when the player spawn, this PR modify two patchs related to handle the world spawn for add a conditional for call a method used to handle location for "new" players in world.

### CODE REVIEW

**Comment by Doc94** [comment]
NOTE: I prefer only edit the patchs... but maybe is better merge the two patchs in one because the two are related to the spawn player. (but dont know how correctly make this and not destroy git in the process)

also... if can add the label for get a JAR for allow the user in the ticket test, i make basic test using vanilla worlds (me in the end, stop server, delete end and start server again)

**Review by Machine-Maker — CHANGES_REQUESTED** [review]
This only handles when the dimension isn't parsed successfully from the playerdata. You need to add something where the server checks if a world exists with the parsed key. Where the logger warns about spawning in an unknown world

File: patches/server/0384-Move-player-to-spawn-point-if-spawn-in-unloaded-worl.patch [hunk 1]
```diff
@@ -6,8 +6,46 @@ Subject: [PATCH] Move player to spawn point if spawn in unloaded world
...
+             Logger logger = PlayerList.LOGGER;
+ 
+             Objects.requireNonNull(logger);
+-            resourcekey = (ResourceKey) dataresult.resultOrPartial(logger::error).orElse(Level.OVERWORLD);
++            // Paper start - detect when cannot find the dimension
++            //resourcekey = (ResourceKey) dataresult.resultOrPartial(logger::error).orElse(Level.OVERWORLD);
++            Optional<ResourceKey<Level>> optionalResourceKey = dataresult.resultOrPartial(logger::error);
++            needResetSpawnPos = optionalResourceKey.isEmpty();
++            resourcekey = optionalResourceKey.orElse(Level.OVERWORLD);
++            // Paper stop
```
Comment (Warriorrrr) [inline patches/server/0384-Move-player-to-spawn-point-if-spawn-in-unloaded-worl.patch:30]:
```suggestion
+            // Paper end
```

**Review by Warriorrrr — COMMENTED** [review]
needResetSpawnPos should also be set to true if the world isn't found like MM said. https://github.com/PaperMC/Paper/blob/73af2d4cea597e65dd13dafab6325ea5709c9067/patches/server/0435-Spawn-player-in-correct-world-on-login.patch also defaults the level key to overworld which will also need to be accounted for.

**Comment by Machine-Maker** [comment]
Superseded by #9887

**Pull request closed without merging**